test/integration/core/node-operators-flow.integration.ts

```diff
@@ -0,0 +1,347 @@
+  );
+  expect(nodeOperatorAfter.totalDepositedValidators).to.equal(nodeOperatorBefore.totalDepositedValidators);
+  expect(nodeOperatorAfter.name).to.equal(nodeOperatorBefore.name);
+  expect(nodeOperatorAfter.rewardAddress).to.equal(nodeOperatorBefore.rewardAddress);
+  expect(nodeOperatorAfter.active).to.equal(expectedChanges.active ?? nodeOperatorBefore.active);
+  if (expectedChanges.totalVettedValidators !== undefined) {
+    expect(nodeOperatorAfter.totalVettedValidators).to.equal(expectedChanges.totalVettedValidators);
+  } else {
+    expect(nodeOperatorAfter.totalVettedValidators).to.equal(nodeOperatorBefore.totalVettedValidators);
+  }
+}
+
+function verifyNodeOperatorSummaryStateChanges(
+  summaryAfter: NodeOperatorSummary,
+  summaryBefore: NodeOperatorSummary,
+  expectedChanges: Partial<NodeOperatorSummary> = {},
+) {
+  expect(summaryAfter.targetLimitMode).to.equal(summaryBefore.targetLimitMode);
+  expect(summaryAfter.targetValidatorsCount).to.equal(summaryBefore.targetValidatorsCount);
+  expect(summaryAfter.stuckValidatorsCount).to.equal(summaryBefore.stuckValidatorsCount);
+  expect(summaryAfter.refundedValidatorsCount).to.equal(summaryBefore.refundedValidatorsCount);
+  expect(summaryAfter.stuckPenaltyEndTimestamp).to.equal(summaryBefore.stuckPenaltyEndTimestamp);
+  expect(summaryAfter.totalExitedValidators).to.equal(summaryBefore.totalExitedValidators);
+  expect(summaryAfter.totalDepositedValidators).to.equal(summaryBefore.totalDepositedValidators);
+  if (expectedChanges.depositableValidatorsCount !== undefined) {
+    expect(summaryAfter.depositableValidatorsCount).to.equal(expectedChanges.depositableValidatorsCount);
+  } else {
+    expect(summaryAfter.depositableValidatorsCount).to.equal(summaryBefore.depositableValidatorsCount);
+  }
+}
+
+describe("Integration: Node operators flow", () => {
+  let ctx: ProtocolContext;
+
+  let snapshot: string;
+  let originalState: string;
+
+  before(async () => {
+    ctx = await getProtocolContext();
+
+    snapshot = await Snapshot.take();
+
+    // Grant required roles
+    const votingSigner = await ctx.getSigner("voting");
+    const agentSigner = await ctx.getSigner("agent");
+
+    await ctx.contracts.stakingRouter
+      .connect(agentSigner)
+      .grantRole(
+        await ctx.contracts.stakingRouter.MANAGE_WITHDRAWAL_CREDENTIALS_ROLE(),
+        await votingSigner.getAddress(),
+      );
+
+    await ctx.contracts.acl
+      .connect(votingSigner)
+      .grantPermission(
+        await votingSigner.getAddress(),
+        ctx.contracts.nor.getAddress(),
+        ethers.keccak256(ethers.toUtf8Bytes("MANAGE_NODE_OPERATOR_ROLE")),
+      );
+  });
+
+  beforeEach(async () => (originalState = await Snapshot.take()));
+
+  afterEach(async () => await Snapshot.restore(originalState));
+
+  after(async () => await Snapshot.restore(snapshot));
+
+  it("Should allow adding a node operator", async () => {
```

Comment:
Maybe split into several logical parts and convert to scenario test, to make it simple to understand what has blown up

Reply:
done